### PR TITLE
update the volume freeze feature to make it unreadable when the volume is frozen

### LINKED ISSUE #2922: [Feature]:To avoid risk, freeze the volume when performing volume deletion and delete it asynchronously after a certain period of time
### Contact Details

_No response_

### Is there an existing issue for this?

- [X] I have searched all the existing issues

### Is your feature request related to a problem? Please describe.

no

### Describe the solution you'd like.

_No response_

### Describe an alternate solution.

_No response_

### Anything else? (Additional Context)

_No response_

### PR DESCRIPTION
**Pull request by shuqiang-zheng**
**What this PR does / why we need it**:
update the volume freeze feature to make it unreadable when the volume is frozen
**Which issue this PR fixes**:

fixes #2922 

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: master/cluster.go [hunk 1]
```diff
@@ -1227,7 +1227,7 @@ func (c *Cluster) deleteVol(name string) {
 	return
 }
 
-func (c *Cluster) markDeleteVol(name, authKey string, force bool) (err error) {
+func (c *Cluster) markDeleteVol(name, authKey string, force bool, status bool) (err error) {
```
Comment (Victor1319) [inline master/cluster.go:1230]: it's better to refactor `status` variable name.

File: master/api_service.go [hunk 2]
```diff
@@ -1814,58 +1814,93 @@ func (m *Server) markDeleteVol(w http.ResponseWriter, r *http.Request) {
...
+				m.cluster.deleteVolMutex.RUnlock()
+				if index == len(m.cluster.delayDeleteVolsInfo)-1 && value.volName != name {
+					msg := fmt.Sprintf("vol[%v] was not previously deleted or already deleted", name)
+					err = errors.New(msg)
+					sendErrReply(w, r, &proto.HTTPReply{Code: proto.ErrCodeVolNotDelete, Msg: err.Error()})
+					return
+				}
+			}
+			m.cluster.deleteVolMutex.Lock()
+			m.cluster.delayDeleteVolsInfo = append(m.cluster.delayDeleteVolsInfo[:index], m.cluster.delayDeleteVolsInfo[index+1:]...)
```
Comment (Victor1319) [inline master/api_service.go:1883]: do you persist this vairable `delayDeleteVolsInfo` ?

File: master/cluster.go (hunk 1 above)
Comment (shuqiang-zheng) [reply master/cluster.go]: Changed variable name to isdelete.

File: master/api_service.go (hunk 2 above)
Comment (shuqiang-zheng) [reply master/api_service.go]: When master executes loadVols(), it makes a judgment based on the forbiden and status of the volume and adds it to delayDeleteVolsInfo based on the result

File: master/cluster.go [hunk 3]
```diff
@@ -1227,7 +1227,7 @@ func (c *Cluster) deleteVol(name string) {
 	return
 }
 
-func (c *Cluster) markDeleteVol(name, authKey string, force bool) (err error) {
+func (c *Cluster) markDeleteVol(name, authKey string, force bool, isdelete bool) (err error) {
```
Comment (leonrayang) [inline master/cluster.go:1230]: isdelete->IsNotCancel

File: master/cluster.go [hunk 4]
```diff
@@ -1238,40 +1238,53 @@ func (c *Cluster) markDeleteVol(name, authKey string, force bool) (err error) {
...
-		volDentryCount := uint64(0)
-		mpsCopy := vol.cloneMetaPartitionMap()
-		for _, mp := range mpsCopy {
-			// to avoid latency, fetch latest mp dentry count from metanode
-			c.doLoadMetaPartition(mp)
-			mpDentryCount := uint64(0)
-			for _, response := range mp.LoadResponse {
-				if response.DentryCount > mpDentryCount {
-					mpDentryCount = response.DentryCount
+	if isdelete {
```
Comment (leonrayang) [inline master/cluster.go:1241]: `
if !isdelete {
	serverAuthKey = vol.Owner
		if !matchKey(serverAuthKey, authKey) {
			return proto.ErrVolAuthKeyNotMatch
		}

		vol.Status = normal
		if err = c.syncUpdateVol(vol); err != nil {
			vol.Status = markDelete
			return proto.ErrPersistenceByRaft
		}
return
}
`

File: master/api_service.go [hunk 5]
```diff
@@ -1814,58 +1814,100 @@ func (m *Server) markDeleteVol(w http.ResponseWriter, r *http.Request) {
...
 			sendErrReply(w, r, newErrHTTPReply(err))
 			return
 		}
-		vol.setDpRdOnly()
-		vol.setMpRdOnly()
-		m.cluster.delayDeleteVolsInfo = append(m.cluster.delayDeleteVolsInfo, &delayDeleteVolInfo{volName: name, authKey: authKey, execTime: time.Now().Add(time.Duration(m.config.volDelayDeleteTime) * time.Hour), user: m.user})
-		log.LogDebugf("delete vol[%v], slice[%v]", name, m.cluster.delayDeleteVolsInfo)
-		msg = fmt.Sprintf("delete vol: forbid vol[%v] successfully,from[%v]", name, r.RemoteAddr)
+		msg = fmt.Sprintf("delete vol[%v] successfully,from[%v]", name, r.RemoteAddr)
 		log.LogWarn(msg)
```
Comment (leonrayang) [inline master/api_service.go:1828]: return directly，else is not needed

File: master/cluster.go (hunk 3 above)
Comment (shuqiang-zheng) [reply master/cluster.go]: Changed variable name to isNotCancel.

File: master/cluster.go (hunk 4 above)
Comment (shuqiang-zheng) [reply master/cluster.go]: Have removed the else and adjusted the code logic.

File: master/api_service.go (hunk 5 above)
Comment (shuqiang-zheng) [reply master/api_service.go:1828]: Have removed the else.

**Review by leonrayang — APPROVED** [review]
LGMT

**Review by Victor1319 — APPROVED** [review]
LGTM

**Pull request merged**